apps/src/p5lab/spritelab/CoreLibrary.js

```diff
@@ -366,12 +369,37 @@ export default class CoreLibrary {
     return speechBubbles[speechBubbles.length - 1];
   }
 
+  reachedSpriteMax() {
+    return this.getNumberOfSprites() >= MAX_NUM_SPRITES;
+  }
+
+  // This function is called within the addSprite function BEFORE a new sprite is created
+  // If the total number of sprites is equal to SPRITE_LIMIT_WARNING, a workspace
+  // alert warning is displayed to let user know they have reached the sprite limit
+  // Note that SPRITE_LIMIT_WARNING = MAX_NUM_SPRITES - 1
+  warnIfAtSpriteLimit() {
+    if (this.getNumberOfSprites() === SPRITE_LIMIT_WARNING) {
+      getStore().dispatch(
+        displayWorkspaceAlert(
+          workspaceAlertTypes.warning,
+          /* display warning when user exceeds SPRITE_LIMIT_WARNING */
+          msg.spriteLimitReached({limit: MAX_NUM_SPRITES}),
+          /* bottom */ true
+        )
+      );
+    }
+  }
+
   /**
    * Adds the specified sprite to the native sprite map
    * @param {Sprite} sprite
    * @returns {Number} A unique id to reference the sprite.
    */
   addSprite(opts) {
+    if (this.reachedSpriteMax()) {
+      return;
+    }
+    this.warnIfAtSpriteLimit();
```

Comment:
NIT: (I don't feel strongly about this. Feel free to ignore if you disagree.) Is `reachedSpriteMax` and `warnIfAtSpriteLimit` linked? i.e. you'd generally expect that the two commands should come one after the other? If so, I'd recommend something like this:
```
    if (this.reachedSpriteMax()) {
      return;
    } else if (this.atSpriteLimit()) {
      this.warn();
    }
```
The reasoning: future code maintainers will read an `if` clause followed by a function as two separate logical blocks but will read an `if`/`else if` clause as one logical block. The two are less likely to be separated and it's less likely that one will be deleted while the other retained.

Reply:
Thank you! I appreciate your point about future code maintainers and added a helper function similar to reachedSpriteMax called reachedSpriteWarningThreshold. Also renamed the function that dispatches the warning - I think it does make it easier to read with the parallel logic:
    if (this.reachedSpriteMax()) {
      return;
    } else if (this.reachedSpriteWarningThreshold()) {
      this.dispatchSpriteLimitWarning();
    }